File: apps/checker/pkg/job/monitors.go
package job

type Monitor struct {
	ID            int         `json:"id"`
	Name          string      `json:"name"`
	URL           string      `json:"url"`
	Periodicity   string      `json:"periodicity"`
	Description   string      `json:"description"`
	Method        string      `json:"method"`
	Regions       []string    `json:"regions"`
	Active        bool        `json:"active"`
	Public        bool        `json:"public"`
	Timeout       int         `json:"timeout"`
	DegradedAfter int         `json:"degraded_after,omitempty"`
	Body          string      `json:"body"`
	Headers       []Header    `json:"headers,omitempty"`
	Assertions    []Assertion `json:"assertions,omitempty"`
	Retry         int         `json:"retry"`
	JobType       string      `json:"jobType"`
}

type Header struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

type Assertion struct {
	Type    string `json:"type"`
	Compare string `json:"compare"`
	Key     string `json:"key"`
	Target  any    `json:"target"`
}

type Timing struct {
	DnsStart          int64 `json:"dnsStart"`
	DnsDone           int64 `json:"dnsDone"`
	ConnectStart      int64 `json:"connectStart"`
	ConnectDone       int64 `json:"connectDone"`
	TlsHandshakeStart int64 `json:"tlsHandshakeStart"`
	TlsHandshakeDone  int64 `json:"tlsHandshakeDone"`
	FirstByteStart    int64 `json:"firstByteStart"`
	FirstByteDone     int64 `json:"firstByteDone"`
	TransferStart     int64 `json:"transferStart"`
	TransferDone      int64 `json:"transferDone"`
}

File: apps/server/src/routes/rpc/services/monitor/converters/monitors.ts
import type { Monitor } from "@openstatus/db/src/schema/monitors/validation";
import type {
  DNSMonitor,
  HTTPMonitor,
  TCPMonitor,
} from "@openstatus/proto/monitor/v1";

import { parseDnsAssertions, parseHttpAssertions } from "./assertions";
import { MONITOR_DEFAULTS } from "./defaults";
import {
  stringToHttpMethod,
  stringToMonitorStatus,
  stringToPeriodicity,
} from "./enums";
import { parseOpenTelemetry, toProtoHeaders } from "./headers";
import { stringsToRegions } from "./regions";

/**
 * Transform database HTTP monitor to proto HTTPMonitor.
 */
export function dbMonitorToHttpProto(dbMon: Monitor): HTTPMonitor {
  const assertions = parseHttpAssertions(dbMon.assertions);

  return {
    $typeName: "openstatus.monitor.v1.HTTPMonitor",
    id: String(dbMon.id),
    name: dbMon.name,
    url: dbMon.url,
    periodicity: stringToPeriodicity(dbMon.periodicity),
    method: stringToHttpMethod(dbMon.method),
    body: dbMon.body ?? "",
    timeout: BigInt(dbMon.timeout),
    degradedAt: dbMon.degradedAfter ? BigInt(dbMon.degradedAfter) : undefined,
    retry: BigInt(dbMon.retry ?? MONITOR_DEFAULTS.retry),
    followRedirects: dbMon.followRedirects ?? MONITOR_DEFAULTS.followRedirects,
    headers: toProtoHeaders(dbMon.headers),
    statusCodeAssertions: assertions.statusCodeAssertions,
    bodyAssertions: assertions.bodyAssertions,
    headerAssertions: assertions.headerAssertions,
    description: dbMon.description,
    active: dbMon.active ?? MONITOR_DEFAULTS.active,
    public: dbMon.public ?? MONITOR_DEFAULTS.public,
    regions: stringsToRegions(dbMon.regions),
    openTelemetry: parseOpenTelemetry(dbMon.otelEndpoint, dbMon.otelHeaders),
    status: stringToMonitorStatus(dbMon.status),
  };
}

/**
 * Transform database TCP monitor to proto TCPMonitor.
 */
export function dbMonitorToTcpProto(dbMon: Monitor): TCPMonitor {
  return {
    $typeName: "openstatus.monitor.v1.TCPMonitor",
    id: String(dbMon.id),
    name: dbMon.name,
    uri: dbMon.url,
    periodicity: stringToPeriodicity(dbMon.periodicity),
    timeout: BigInt(dbMon.timeout),
    degradedAt: dbMon.degradedAfter ? BigInt(dbMon.degradedAfter) : undefined,
    retry: BigInt(dbMon.retry ?? MONITOR_DEFAULTS.retry),
    description: dbMon.description,
    active: dbMon.active ?? MONITOR_DEFAULTS.active,
    public: dbMon.public ?? MONITOR_DEFAULTS.public,
    regions: stringsToRegions(dbMon.regions),
    openTelemetry: parseOpenTelemetry(dbMon.otelEndpoint, dbMon.otelHeaders),
    status: stringToMonitorStatus(dbMon.status),
  };
}

/**
 * Transform database DNS monitor to proto DNSMonitor.
 */
export function dbMonitorToDnsProto(dbMon: Monitor): DNSMonitor {
  return {
    $typeName: "openstatus.monitor.v1.DNSMonitor",
    id: String(dbMon.id),
    name: dbMon.name,
    uri: dbMon.url,
    periodicity: stringToPeriodicity(dbMon.periodicity),
    timeout: BigInt(dbMon.timeout),
    degradedAt: dbMon.degradedAfter ? BigInt(dbMon.degradedAfter) : undefined,
    retry: BigInt(dbMon.retry ?? MONITOR_DEFAULTS.retry),
    recordAssertions: parseDnsAssertions(dbMon.assertions),
    description: dbMon.description,
    active: dbMon.active ?? MONITOR_DEFAULTS.active,
    public: dbMon.public ?? MONITOR_DEFAULTS.public,
    regions: stringsToRegions(dbMon.regions),
    openTelemetry: parseOpenTelemetry(dbMon.otelEndpoint, dbMon.otelHeaders),
    status: stringToMonitorStatus(dbMon.status),
  };
}
